### PR TITLE
Seeding RL Folder

### PR DESCRIPTION
**Pull request by harshmahesheka**
Partially adding RL implementations, which was part of my GSoC project, to mesa-examples. The current pr contains the addition of documentation, tutorials and the Epstein-civil violence model.

### CODE REVIEW

File: rl/README.md
Comment (EwoutH) [inline rl/README.md:18]: I think this can just be `pip install mesa<3`

(maybe quotes are needed)

File: rl/README.md
Comment (EwoutH) [inline rl/README.md:43]: Could you be a little more specific here (if possible)?

File: rl/README.md
Comment (EwoutH) [inline rl/README.md:61]: I like this section, but I think this last paragraph isn’t needed.

(also I get triggered by the word invaluable, maybe because ChatGPT uses it a lot)

**Review by EwoutH — COMMENTED** [review]
Thanks a lot for the PR! I will try to review it in detail Wednesday.

On a high level, the way you structured it looks good!

File: rl/README.md
Comment (rht) [inline rl/README.md:3]: s/my/<your name>/ because the context (where the repo is originally located) is gone.

File: rl/model/boltzmann_money.zip
Comment (rht) [inline rl/model/boltzmann_money.zip:1]: What is the content of this (relatively) huge zip file?

File: rl/model/epstein_civil_violence/policies/policy_citizen/policy_state.pkl
Comment (rht) [inline rl/model/epstein_civil_violence/policies/policy_citizen/policy_state.pkl:1]: Should we include this file in the repo? This is 2 MB. I prefer not, but other maintainers might have a reasonable take to include it. How long does it take to train to get to this file?

Comment (rht) [reply rl/model/epstein_civil_violence/policies/policy_citizen/policy_state.pkl:1]: In total they add up to ~13 MB.

Comment (rht) [reply rl/model/epstein_civil_violence/policies/policy_citizen/policy_state.pkl:1]: Is there a somewhat permanent place for these files? Maybe Hugging Face?

File: rl/README.md
Comment (harshmahesheka) [reply rl/README.md:61]: Will remove it

File: rl/model/epstein_civil_violence/policies/policy_citizen/policy_state.pkl
Comment (harshmahesheka) [reply rl/model/epstein_civil_violence/policies/policy_citizen/policy_state.pkl:1]: We can do that; it would make sense. We can have a mesa id where it will be permanent

File: rl/model/boltzmann_money.zip
Comment (harshmahesheka) [reply rl/model/boltzmann_money.zip:1]: It's just weighted for trained policy for the Boltzman environment.

File: rl/model/epstein_civil_violence/policies/policy_citizen/policy_state.pkl
Comment (rht) [reply rl/model/epstein_civil_violence/policies/policy_citizen/policy_state.pkl:1]: cc: @jackiekazil @tpike3 we need a Hugging Face repo to store these.

File: rl/README.md
Comment (rht) [reply rl/README.md:18]: Quotes are needed for Zsh (which is the default shell on macOS).

File: rl/model/epstein_civil_violence/policies/policy_citizen/policy_state.pkl
Comment (EwoutH) [reply rl/model/epstein_civil_violence/policies/policy_citizen/policy_state.pkl:1]: I think we can just make one and give all maintainers access.

Comment (EwoutH) [reply rl/model/epstein_civil_violence/policies/policy_citizen/policy_state.pkl:1]: @harshmahesheka Sorry for the wait. I created an organization on HuggingFace: https://huggingface.co/projectmesa

I send you an invite link on Element.

File: rl/README.md
Comment (harshmahesheka) [reply rl/README.md:3]: Removed the GSoC bit was more suited for my repo

File: rl/README.md
Comment (harshmahesheka) [reply rl/README.md:18]: Can we source mesa_models just installing mesa?

File: rl/README.md
Comment (harshmahesheka) [reply rl/README.md:43]: Changed the wording to make it clearer

File: rl/README.md
Comment (EwoutH) [reply rl/README.md:18]: I was wrong, sorry, disregard this comment. You did it the right way.

File: rl/epstein_civil_violence/model.py
Comment (EwoutH) [inline rl/epstein_civil_violence/model.py:68]: Add for both observation_space and action_space comments explaining what they are and what the user should know.

File: rl/epstein_civil_violence/model.py
Comment (EwoutH) [inline rl/epstein_civil_violence/model.py:90]: Was is a conscious decision to collect data before the remainder of the step was performed?

File: rl/epstein_civil_violence/model.py
Comment (EwoutH) [inline rl/epstein_civil_violence/model.py:105]: I see already a place where the brand new [AgentSet `set` method](https://github.com/projectmesa/mesa/pull/2254) can be useful!

File: rl/epstein_civil_violence/model.py
Comment (EwoutH) [inline rl/epstein_civil_violence/model.py:167]: This seems a bit weird. Can't we call init again or something like that?

File: rl/epstein_civil_violence/resources/epstein.gif
Comment (EwoutH) [inline rl/epstein_civil_violence/resources/epstein.gif:1]: Very nice visualisation!

Can we add some metric for how well the cops are doing?

File: rl/epstein_civil_violence/train_config.py
Comment (EwoutH) [inline rl/epstein_civil_violence/train_config.py:1]: Add some docstring/comments explaining what's happening here.

File: rl/epstein_civil_violence/utility.py
Comment (EwoutH) [inline rl/epstein_civil_violence/utility.py:75]: Would using a dict be elegant here?
```suggestion
def move(self, action, empty_neighbors):
    # Define the movement deltas
    moves = {
        0: (1, 0),   # Move right
        1: (-1, 0),  # Move left
        2: (0, -1),  # Move up
        3: (0, 1),   # Move down
    }
    
    # Get the delta for the action, defaulting to (0, 0) if the action is invalid
    dx, dy = moves.get(int(action), (0, 0))
    
    # Calculate the new position and wrap around the grid
    new_position = (
        (self.pos[0] + dx) % self.model.grid.width,
        (self.pos[1] + dy) % self.model.grid.height
    )
    
    # Move the agent if the new position is in empty_neighbors
    if new_position in empty_neighbors:
        self.model.grid.move_agent(self, new_position)
```

File: rl/train.py
Comment (EwoutH) [inline rl/train.py:1]: Add some more docstring and comments

**Review by EwoutH — COMMENTED** [review]
Great work, it looks like a very complete PR including documentation and visualisation. I added some comments. They are all just suggestions, if you like them you can adopt them if not it's also okay.

Finally, consider adding a `test_rl` optional dependency in the `pyproject.toml`, so we know what dependencies are needed to run this RL example.

https://github.com/projectmesa/mesa-examples/blob/3d29e4c215154396e20a5a39a09b8a4043b22de6/pyproject.toml#L12-L20

File: rl/model/epstein_civil_violence/policies/policy_citizen/policy_state.pkl
Comment (harshmahesheka) [reply rl/model/epstein_civil_violence/policies/policy_citizen/policy_state.pkl:1]: Done

File: rl/Tutorials.md
Comment (harshmahesheka) [reply rl/Tutorials.md:1]: Done

File: rl/epstein_civil_violence/agent.py
Comment (harshmahesheka) [reply rl/epstein_civil_violence/agent.py:10]: I think any value other than 0 is interpreted as True

File: rl/epstein_civil_violence/agent.py
Comment (harshmahesheka) [reply rl/epstein_civil_violence/agent.py:29]: It is intended one time we are updating before moving it to new spot to get correct empty spaces around, second time we are updating it to get correct observation after movement

File: rl/epstein_civil_violence/agent.py
Comment (harshmahesheka) [reply rl/epstein_civil_violence/agent.py:40]: Done

File: rl/epstein_civil_violence/agent.py
Comment (harshmahesheka) [reply rl/epstein_civil_violence/agent.py:42]: As I mentioned above, the logic is True for all cases other than 0

File: rl/epstein_civil_violence/agent.py
Comment (harshmahesheka) [reply rl/epstein_civil_violence/agent.py:46]: Done

File: rl/epstein_civil_violence/agent.py
Comment (harshmahesheka) [reply rl/epstein_civil_violence/agent.py:17]: It is action dict beacuse I am sotring it as model variable. I have upated the code to make it more readable

Comment (EwoutH) [reply rl/epstein_civil_violence/agent.py:17]: Thanks!

File: rl/epstein_civil_violence/model.py
Comment (harshmahesheka) [reply rl/epstein_civil_violence/model.py:90]: Yes, the agent moved and we collect data

File: rl/epstein_civil_violence/agent.py
Comment (EwoutH) [reply rl/epstein_civil_violence/agent.py:10]: > Python considers any non-zero number as `True` and zero as `False`.

You're actually right. Maybe make it like this to make it fully robust and explicit.

```suggestion
        # If in jail decrease sentence, else update condition
        if self.jail_sentence > 0:
            self.jail_sentence -= 1
```

File: rl/epstein_civil_violence/model.py
Comment (harshmahesheka) [reply rl/epstein_civil_violence/model.py:105]: Is it working with Mesa 2?

File: rl/epstein_civil_violence/model.py
Comment (harshmahesheka) [reply rl/epstein_civil_violence/model.py:167]: The problem is with init we call Citizen class whereas we want CItizenRL class hence the need to reimplement it

File: rl/epstein_civil_violence/model.py
Comment (EwoutH) [reply rl/epstein_civil_violence/model.py:105]: As of about three hours ago ;)

https://github.com/projectmesa/mesa/releases/tag/v2.3.3

Comment (EwoutH) [reply rl/epstein_civil_violence/model.py:105]: But you're creating a dict and not setting a variable, so I made a wrong suggestion.

File: rl/epstein_civil_violence/resources/epstein.gif
Comment (harshmahesheka) [reply rl/epstein_civil_violence/resources/epstein.gif:1]: I have added some lines about the nature of Readme; specifying particular metrics might be challenging.

File: rl/epstein_civil_violence/train_config.py
Comment (harshmahesheka) [reply rl/epstein_civil_violence/train_config.py:1]: Done

File: rl/epstein_civil_violence/utility.py
Comment (harshmahesheka) [reply rl/epstein_civil_violence/utility.py:75]: Sure, changed

File: rl/train.py
Comment (harshmahesheka) [reply rl/train.py:1]: Added

**Comment by harshmahesheka** [comment]
Thanks for all the suggestions, @EwoutH and @rht . I have included most of that suggestions and pushed updated code.

File: rl/epstein_civil_violence/agent.py
Comment (harshmahesheka) [reply rl/epstein_civil_violence/agent.py]: Done

File: rl/epstein_civil_violence/resources/epstein.gif
Comment (EwoutH) [reply rl/epstein_civil_violence/resources/epstein.gif:1]: Something like mean arrest per cop per day?

File: pyproject.toml
Comment (EwoutH) [inline pyproject.toml]: This is purely for the tests in CI. I think that should be:

```suggestion
rl_test = [
    "pytest",
    "stable-baselines3",
    "tensorboard"
]
```
Mesa is separately installed and seaborn is probably not used in the model run itself.

Which file should we run if we want to properly test this code in CI? We will be stepping the model a few times, but do we also need to test/run other files?

File: pyproject.toml
Comment (EwoutH) [inline pyproject.toml:25]: This is purely for the tests in CI. I think that should be:

```suggestion
test_rl = [
    "pytest",
    "stable-baselines3",
    "tensorboard",
]
```
Mesa is separately installed and seaborn is probably not used in the model run itself.

Which file should we run if we want to properly test this code in CI? We will be stepping the model a few times, but do we also need to test/run other files?

Comment (harshmahesheka) [reply pyproject.toml:25]: I think RLlib requires seaborn. Hence, it must be installed. Also, if example.py runs, everything works

**Comment by EwoutH** [comment]
I'm doing some CI testing on [this branch](https://github.com/EwoutH/mesa-examples/tree/rl_ci). CI runs [here](https://github.com/EwoutH/mesa-examples/actions?query=branch%3Arl_ci).

File: rl/rl_models
Comment (EwoutH) [inline rl/rl_models:1]: This submodule is very heavy. I don't know if this is any benefit to doing it this way. Can we dynamically load the models where we need them?

CC @rht

**Comment by EwoutH** [comment]
I seem to be needing far more dependencies, at least

```toml
rl_example = [
    "stable-baselines3",
    "seaborn",
    "mesa",
    "tensorboard",
    "ray",
    "dm_tree",
    "pyarrow",
    "typer",
]
```

**Comment by EwoutH** [comment]
One more point: Could you place everything specific to the `Epstein` model in an separate folder? Most already is, just make sure everything in the root `rl` folder is non-Epstein specific.

After that I'm good to merge, we could improve other points later.

**Comment by harshmahesheka** [comment]
> One more point: Could you place everything specific to the `Epstein` model in an separate folder? Most already is, just make sure everything in the root `rl` folder is non-Epstein specific.
> 
> After that I'm good to merge, we could improve other points later.

I think everything related to Epstein is in Esptein's folder. The example.py shows how to run esptein's model hence, it is in rl folder. It will be common for all runs

File: rl/rl_models
Comment (EwoutH) [reply rl/rl_models:1]: I think this is okay for now.

**Review by EwoutH — APPROVED** [review]
Okay, we probably will encounter some stuff when we're adding the CI, but we can solve those in separate PRs. For now this looks good.

Shall I merge?

**Comment by EwoutH** [comment]
@harshmahesheka for context, I asked @jackiekazil for a review, since she was your primary mentor on this project.

**Comment by jackiekazil** [comment]
This looks good - I was already tracking. Thank you for this addition! 
Between @EwoutH and @rht's review, I think everything was covered.

**Review by jackiekazil — APPROVED** [review]

**Pull request merged**
